app/gui2/src/util/ffi.ts

```diff
@@ -0,0 +1,406 @@
+    }
+    export interface Function {
+      name: Tree
+      args: ArgumentDefinition[]
+      equals: Token.Operator
+      body: Opt<Tree>
+    }
+    export interface ForeignFunction {
+      foreign: Token.Ident
+      language: Token.Ident
+      name: Token.Ident
+      args: ArgumentDefinition[]
+      equals: Token.Operator
+      body: Tree
+    }
+    export interface Import {
+      polyglot: Opt<MultiSegmentAppSegment>
+      from: Opt<MultiSegmentAppSegment>
+      import: MultiSegmentAppSegment
+      all: Opt<Token.Ident>
+      as_: Opt<MultiSegmentAppSegment>
+      hiding: Opt<MultiSegmentAppSegment>
+    }
+    export interface Export {
+      from: Opt<MultiSegmentAppSegment>
+      export: MultiSegmentAppSegment
+      all: Opt<Token.Ident>
+      as_: Opt<MultiSegmentAppSegment>
+      hiding: Opt<MultiSegmentAppSegment>
+    }
+    export interface Group {
+      open: Opt<Token.OpenSymbol>
+      body: Opt<Tree>
+      close: Opt<Token.CloseSymbol>
+    }
+    export interface TypeSignature {
+      variable: Tree
+      operator: Token.Operator
+      type_: Tree
+    }
+    export interface TypeAnnotated {
+      expression: Tree
+      operator: Token.Operator
+      type_: Tree
+    }
+
+    export interface CaseLine {
+      newline: Opt<Token.Newline>
+      case: Opt<Case>
+    }
+
+    export interface Case {
+      documentation: Opt<DocComment>
+      pattern: Opt<Tree>
+      arrow: Opt<Token.Operator>
+      expression: Opt<Tree>
+    }
+
+    export interface CaseOf {
+      case: Token.Ident
+      expression: Opt<Tree>
+      of: Token.Ident
+      cases: CaseLine[]
+    }
+    export interface Lambda {
+      operator: Token.Operator
+      arrow: Opt<Tree>
+    }
+
+    export interface OperatorDelimitedTree {
+      operator: Token.Operator
+      body: Opt<Tree>
+    }
+
+    export interface Array {
+      left: Token.OpenSymbol
+      first: Opt<Tree>
+      rest: OperatorDelimitedTree[]
+      right: Token.CloseSymbol
+    }
+    export interface Tuple {
+      left: Token.OpenSymbol
+      first: Opt<Tree>
+      rest: OperatorDelimitedTree[]
+      right: Token.CloseSymbol
+    }
+    export interface Annotated {
+      token: Token.Operator
+      annotation: Token.Ident
+      argument: Opt<Tree>
+      newlines: Token.Newline[]
+      expression: Opt<Tree>
+    }
+    export interface AnnotatedBuiltin {
+      token: Token.Operator
+      annotation: Token.Ident
+      newlines: Token.Newline[]
+      expression: Opt<Tree>
+    }
+    export interface Documented {
+      documentation: DocComment
+      expression: Opt<Tree>
+    }
+    export interface ConstructorDefinition {
+      constructor: Token.Ident
+      arguments: ArgumentDefinition[]
+      block: ArgumentDefinitionLine[]
+    }
+  }
+
+  export interface DocComment {
+    open: Token.TextStart
+    elements: TextElement[]
+    newlines: Token.Newline[]
+  }
+
+  export type TextElement =
+    | { Section: TextElement.Section }
```

Comment:
typescript prefers discriminated unions (so the variants would be encoded as a string inside the property). cloud seems to use `#[serde(tag = "type")]` for that - although the identifier can be whatever is convenient